lightning/src/ln/onion_utils.rs

```diff
@@ -592,4 +612,58 @@ mod tests {
 		let onion_packet_5 = super::encrypt_failure_packet(&onion_keys[0].shared_secret[..], &onion_packet_4.data[..]);
 		assert_eq!(onion_packet_5.data, hex::decode("9c5add3963fc7f6ed7f148623c84134b5647e1306419dbe2174e523fa9e2fbed3a06a19f899145610741c83ad40b7712aefaddec8c6baf7325d92ea4ca4d1df8bce517f7e54554608bf2bd8071a4f52a7a2f7ffbb1413edad81eeea5785aa9d990f2865dc23b4bc3c301a94eec4eabebca66be5cf638f693ec256aec514620cc28ee4a94bd9565bc4d4962b9d3641d4278fb319ed2b84de5b665f307a2db0f7fbb757366067d88c50f7e829138fde4f78d39b5b5802f1b92a8a820865af5cc79f9f30bc3f461c66af95d13e5e1f0381c184572a91dee1c849048a647a1158cf884064deddbf1b0b88dfe2f791428d0ba0f6fb2f04e14081f69165ae66d9297c118f0907705c9c4954a199bae0bb96fad763d690e7daa6cfda59ba7f2c8d11448b604d12d").unwrap());
 	}
+
+	struct RawOnionHopData {
+		data: Vec<u8>
+	}
+	impl RawOnionHopData {
+		fn new(orig: msgs::OnionHopData) -> Self {
+			Self { data: orig.encode() }
+		}
+	}
+	impl Writeable for RawOnionHopData {
+		fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+			writer.write_all(&self.data[..])
+		}
+	}
+
+	#[test]
+	fn tlv_onion_vectors() {
+		// Packet creation test vectors from BOLT 4
+		// Note that we use he RawOnionHopData for everything except Legacy hops, as even the hops
+		// with "type": "tlv" are not valid TLV, and, thus, cannot be directly represented in our
+		// in-memory enums.
```

Comment:
I'm not sure I follow this. Why are we using invalid TLVs in this test? What exactly is the behavior being tested?

Reply:
These are the test vectors from BOLT 4. They are based on a previous version of TLV, so are invalid in today's parlance.

Reply:
Explained to me a bit offline. I think it would be more accurate to call this test something like `variable_length_hop_payload` given it contains legacy format, raw format, and "previous" TLV format. I'd also recommend linking to the test vector file in the comment.